controlplane/pkg/model/model.go

```diff
@@ -143,7 +144,32 @@ func (i *impl) DeleteEndpoint(name string) error {
 	return fmt.Errorf("no endpoint with name: %s", name)
 }
 
-func (i *impl) SelectEndpoint(nsName string) (*registry.NSERegistration, error) {
+func (i *impl) selectEndpointRR(nsName string, registrations *[]*registry.NSERegistration) (*registry.NSERegistration, error) {
+	nseRegistrations := *registrations
+	idx := i.roundRobin[nsName] % len(nseRegistrations)
+	endpoint := nseRegistrations[idx]
+	if endpoint == nil {
+		return nil, errors.New("should not see this error, scaffolding called")
+	}
+	i.roundRobin[nsName] = (idx + 1) % len(nseRegistrations)
+	return endpoint, nil
+}
+
+func (i *impl) selectEndpointMatch(nsName string, nsLabels *map[string]string, registrations *[]*registry.NSERegistration) (*registry.NSERegistration, error) {
```

Comment:
If this becomes:

func (i *impl) selectEndpointMatch(ns *registry.NetworkService, sourceLabels map[string]string,registrations *[]*registry.NSERegistration)

Then you can loop over it something like:

```go
var registration *registry.NSERegistration
var match *registry.Match
// Loop over the matches
for _,match = range ns.GetMatches() {
  // Note: pseudocode, so add in the appropriate safety checks :)
  // Loop over the SourceSelector labels in the match
  for key,value := range match.GetSourceSelector() {
      // If the sourceLabels fail to completely match the sourceSelector, we don't match, and we break from ths
      // sourceSelector loop
      if sourceLabel[key] != value {
        match = nil
        break
      }
  }
  var candidateRegistrations []*registry.Destinations
  // If we still have a match that isn't nil, match.GetRoutes() will give us the list of destinations
  // if match == nil, then this loop is a noop.
  for _,destination := range match.GetRoutes() {
       // Loop through the registrations looking for ones that have matching DestinationLabels
        for _,registration := range registrations {
             isCandidate = true
             for key,value := registration.GetLabels() {
                   // If the registration Labels don't match the DestinationSelector, break and declare it so
                   if destination.GetDestinationSelector()[key] != value {
                       isCandidate = false
                       break
                   }
             }
             if isCandidate {
                candidateRegistrations := append(candidateRegistrations,registration)
             }
        }
    }
    // until we get to LBing by weight, just return the first one:
    return candidateRegistrations[0]
}
```

Reply:
Note: this shows the full logic (I think) don't necessarily do this all in one long nested loop ;)